device/cuda/src/clusterization/clusterization_algorithm.cu

```diff
@@ -89,181 +276,64 @@ clusterization_algorithm::clusterization_algorithm(
-    kernels::find_clusters<<<blocksPerGrid, threadsPerBlock, 0, stream>>>(
-        cells_view, sparse_ccl_indices_buff, cl_per_module_prefix_buff);
     CUDA_ERROR_CHECK(cudaGetLastError());
-
-    // Create prefix sum buffer
-    vecmem::data::vector_buffer cells_prefix_sum_buff =
-        make_prefix_sum_buff(cell_sizes, m_copy, m_mr, m_stream);
-
-    // Copy the sizes of clusters per module to the host
-    // and create a copy of "clusters per module" vector
-    vecmem::vector<std::size_t> cl_per_module_prefix_host(
-        m_mr.host ? m_mr.host : &(m_mr.main));
-    m_copy(cl_per_module_prefix_buff, cl_per_module_prefix_host,
-           vecmem::copy::type::copy_type::device_to_host);
     m_stream.synchronize();
-    std::vector<std::size_t> clusters_per_module_host(
-        cl_per_module_prefix_host.begin(), cl_per_module_prefix_host.end());
-
-    // Perform the inclusive scan operation
-    std::inclusive_scan(cl_per_module_prefix_host.begin(),
-                        cl_per_module_prefix_host.end(),
-                        cl_per_module_prefix_host.begin());
-
-    unsigned int total_clusters = cl_per_module_prefix_host.back();
-
-    // Copy the prefix sum back to its device container
-    m_copy(vecmem::get_data(cl_per_module_prefix_host),
-           cl_per_module_prefix_buff,
-           vecmem::copy::type::copy_type::host_to_device);
-
-    // Vector of the exact cluster sizes, will be filled in cluster counting
-    vecmem::data::vector_buffer<unsigned int> cluster_sizes_buffer(
-        total_clusters, m_mr.main);
-    m_copy.setup(cluster_sizes_buffer);
-    m_copy.memset(cluster_sizes_buffer, 0);
-
-    // Calclating grid size for cluster counting kernel (block size 64)
-    blocksPerGrid = (cells_prefix_sum_buff.capacity() + threadsPerBlock - 1) /
-                    threadsPerBlock;
-    // Invoke cluster counting will call count cluster cells kernel
-    kernels::count_cluster_cells<<<blocksPerGrid, threadsPerBlock, 0, stream>>>(
-        sparse_ccl_indices_buff, cl_per_module_prefix_buff,
-        cells_prefix_sum_buff, cluster_sizes_buffer);
-    // Check for kernel launch errors and Wait for the cluster_counting kernel
-    // to finish
-    CUDA_ERROR_CHECK(cudaGetLastError());
 
-    // Copy cluster sizes back to the host
-    vecmem::vector<unsigned int> cluster_sizes{m_mr.host ? m_mr.host
-                                                         : &(m_mr.main)};
-    m_copy(cluster_sizes_buffer, cluster_sizes,
-           vecmem::copy::type::copy_type::device_to_host);
-    m_stream.synchronize();
+    // Copy number of measurements to host
+    unsigned int num_measurements_host;
+    CUDA_ERROR_CHECK(cudaMemcpy(&num_measurements_host,
+                                num_measurements_device.get(),
+                                sizeof(unsigned int), cudaMemcpyDeviceToHost));
+    // I doubt this is necessary here
+    CUDA_ERROR_CHECK(cudaDeviceSynchronize());
```

Comment:
Just thought there wouldn't really be a big difference between having a synchronous copy or having an assynchronous one immediately followed by a synchronisation point, given that it's just a single 32bit copy. And being synchronous just makes the code slightly more compact.

Reply:
With a synchronous copy you need to:
  - Wait for the kernel that produced the value on the device, before the copy;
  - Schedule the copy, and wait for its results;

If you set it up asynchronously, you can run the CPU code that schedules the copy, while the GPU is still in the process of producing the output. (You don't need to have a synchronization point before the copy.) You do need a synchronization point **after** the copy, since you want to make use of the value copied back to the host. But you will need to do fewer things on the CPU while the GPU is idle. (Since the setup of the copy is done in that case while the GPU is ideally already busy.)